### PR TITLE
VASP - change NSW if interactive

### PR DESCRIPTION
**Pull request by raynol-dsouza**
Modify the `NSW` flag to `1000` if VASP is run interactively.

### CODE REVIEW

**Comment by liamhuber** [comment]
I don't see how the commit does what the title suggests?

**Comment by raynol-dsouza** [comment]
@liamhuber the check for interactive is done in `_run_if_new()`, which in turn calls `_check_incar_parameters()` where the change is made.

**Comment by liamhuber** [comment]
> @liamhuber the check for interactive is done in `_run_if_new()`, which in turn calls `_check_incar_parameters()` where the change is made.

Doesn't the number 1000 still need to appear somewhere, or is this just setting the interactive NSW value to the default (calc static? calc md?) NSW value?

**Comment by raynol-dsouza** [comment]
```
def _run_if_new(self, debug=False):
        if (
            self.server.run_mode.interactive
            or self.server.run_mode.interactive_non_modal
        ):
            self._check_incar_parameter(parameter="INTERACTIVE", value=True)
            self._check_incar_parameter(parameter="IBRION", value=-1)
            self._check_incar_parameter(parameter="POTIM", value=0.0)
            self._check_incar_parameter(parameter="NSW", value=1000)
            self._check_incar_parameter(parameter="ISYM", value=0)
        super(VaspInteractive, self)._run_if_new(debug=debug)
```

I believe it is a hard coded value.

**Comment by liamhuber** [comment]
Hmmm, I have multiple issues here.

If I understand correctly, when the NSW value is not large enough the VASP process terminates before `interactive_close` can be called and the interactive job is not able to collect output data correctly. Currently, the interactive job sets NSW to 1000 *if it is not yet set*. Here, your goal is to *always override* this value with 1000 whether it's set or not.

What I don't like about your solution is that the correction is buried as deep as possible, all the way down in `_check_incar_parameter`. Definitely this can get bumped up to `_run_if_new` by replacing `self._check_incar_parameter(parameter="NSW", value=1000)` with `self.input.incar['NSW'] = 1000`. This has the same effect, but is less opaque. Since a large NSW value is a prerequisite for running VASP interactively, I would like it even better if this line was in `validate_ready_to_run`, since validation of readiness is exactly what this is, but I'm not 100% sure that putting it there won't introduce some bug.

What I do like about your solution is it really moves towards forcing the interactive run to work, regardless of what silly input the user has provided. But then I'm not such a fan of the original code -- why `1000`? What if the user wants to do more than 1000 iterations? An Al unit cell takes only 2s to `calc_static` on the head node, so this is not inconcievable. The current code fixes the problem only in heuristic way, and exposes the user to a surprising bug if (a) they override the NSW value (which you now fix) or (b) if they run for longer than the devs expected (which still remains after this PR).

If jacking up the NSW value to some large integer is really ok, why not go all the way? `self.input.incar['NSW'] = np.iinfo(np.int32).max - 1` just sets it to one less the biggest possible (signed) int32. Putting that as high up the interactive food chain as possible solves the problem completely and with less ambiguity of purpose.

(Note: I originally tried this manually with int64, but that breaks things. The error message was unclear to me, but it obviously must have been some overflow. Dropping to int32 worked fine. I don't think we need to worry anymore about people running pyiron on a 16 bit system, nor about people running more than 2147483646 iterations since that works out to 136 years even with the 2s iteration for an Al one-atom unit cell...)

**Comment by liamhuber** [comment]
This might feel like overkill, but if we're going to squash this bug let's squash it forever! 😈 

only good bug

**Comment by raynol-dsouza** [comment]
Although I do agree with you, @liamhuber, since this is my first time trying to get some jobs run using VASP, I would rather have people who use VASP more frequently take a call on this.

**Comment by liamhuber** [comment]
After looking at your issue page, I remembered that both Dominik and Hari ran into exactly this problem with QM/MM, so I'm glad you took the time to figure out where it needs to be addressed!

**Comment by raynol-dsouza** [comment]
@liamhuber I only ran into the bug! It was @jan-janssen who isolated where the bug was. :)

**Comment by liamhuber** [comment]
> @liamhuber I only ran into the bug! It was @jan-janssen who isolated where the bug was. :)

No no Raynol, you're supposed to take credit for your excellent management work in delegating the problem 😆 

But seriously, I'm glad you guys tracked it down, regardless who did the bug finding and who did the line finding 👍

**Comment by raynol-dsouza** [comment]
@liamhuber should this PR be merged? It has already been merged with `vasp_int_conv_check`, which has been merged with master.

**Comment by liamhuber** [comment]
> @liamhuber should this PR be merged? It has already been merged with `vasp_int_conv_check`, which has been merged with master.

The `convergence_check` code is definitely conflicting with #222, which should be merged in here. Learning from that PR, my current understand is that you also need a `if self.server.run_mode.interactive:` catch around your NSW modification here, or? At the same time, instead of 1000 let's really use a properly big number (`np.iinfo(np.int32).max - 1`) so we don't need to come re-fix it later.

Ultimately you should probably also get a review from someone who knows the VASP code. The PR itself is only a couple of lines, it's just a matter of being familiar with other pieces it might interact with. @sudarsan-surendralal?

File: pyiron_atomistics/vasp/interactive.py [hunk 1]
```diff
@@ -296,6 +299,16 @@ def _run_if_new(self, debug=False):
             self._check_incar_parameter(parameter="ISYM", value=0)
         super(VaspInteractive, self)._run_if_new(debug=debug)
 
+    def convergence_check(self):
```
Comment (sudarsan-surendralal) [inline pyiron_atomistics/vasp/interactive.py]: Merge the master branch into this which updates this function!

**Review by sudarsan-surendralal — CHANGES_REQUESTED** [review]
I agree with @liamhuber, it's better to use something like `np.iinfo(np.int32).max - 1` in `_run_if_new` rather than limiting it to something like 1000.

File: pyiron_atomistics/vasp/interactive.py (hunk 1 above)
Comment (raynol-dsouza) [reply pyiron_atomistics/vasp/interactive.py]: Done.

**Review by sudarsan-surendralal — APPROVED** [review]
Looks good to me! Feel free to merge this @raynol-dsouza

**Pull request merged**